### PR TITLE
feat(utxo-lib): refactor psbt input updates out of add

### PR DESCRIPTION
**Pull request by davidkaplanbitgo**
By refactoring out the input updates out from adding the inputs to the PSBT, we are able to add missing information to a PSBT if the one we are given is incomplete.

In addition, there is a fix to adding Musig2 nonces so that they are at the right input.

BG-72538

### CODE REVIEW

File: modules/utxo-lib/src/bitgo/wallet/Unspent.ts [hunk 1]
```diff
@@ -246,3 +302,18 @@ export function addWalletUnspentToPsbt(
...
+  rootWalletKeys: RootWalletKeys,
+  signer: KeyName,
+  cosigner: KeyName,
+  /**
+   * @deprecated
+   */
+  network: Network = psbt.network
+): void {
+  addBaseUnspentToPsbt(psbt, u, network);
+  updateWalletUnspentForPsbt(psbt, psbt.inputCount - 1, u, rootWalletKeys, signer, cosigner);
```
Comment (saravanan7mani) [inline modules/utxo-lib/src/bitgo/wallet/Unspent.ts:318]: Does PSBT object allows updating if data already exist?

File: modules/utxo-lib/src/bitgo/wallet/Unspent.ts [hunk 2]
```diff
@@ -120,57 +120,113 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
...
+ * @param inputIndex
+ * @param u
+ * @returns true iff the unspent ID on the unspent and psbt input match
+ */
+export function unspentCorrespondsToPsbtInput(
   psbt: UtxoPsbt,
+  inputIndex: number,
+  u: Unspent<bigint> | WalletUnspent<bigint>
+): boolean {
+  if (inputIndex < 0 || inputIndex >= psbt.inputCount) {
```
Comment (saravanan7mani) [inline modules/utxo-lib/src/bitgo/wallet/Unspent.ts:134]: `checkForInput(psbt.data.inputs, inputIndex)` can be used.

**Review by saravanan7mani — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/src/bitgo/wallet/Unspent.ts [hunk 3]
```diff
@@ -120,129 +122,220 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
...
   }
+  if (isZcash && !input.witnessUtxo) {
+    const { script, value } = toPrevOutput(u, psbt.network);
+    psbt.updateInput(psbt.inputCount - 1, { witnessUtxo: { script, value } });
+  } else if (!isZcash && !input.nonWitnessUtxo) {
+    psbt.updateInput(psbt.inputCount - 1, { nonWitnessUtxo: (u as UnspentWithPrevTx<bigint>).prevTx });
+  }
+}
+
+function addBaseUnspentToPsbt(psbt: UtxoPsbt, u: Unspent<bigint>): void {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/wallet/Unspent.ts:182]:
```suggestion
function addUnspentToPsbt(psbt: UtxoPsbt, u: Unspent<bigint>): void {
```

File: modules/utxo-lib/src/bitgo/wallet/Unspent.ts [hunk 4]
```diff
@@ -120,129 +122,220 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
...
+}
+
+/**
+ * Update the psbt input at the given index
+ * @param psbt
+ * @param inputIndex
+ * @param u
+ * @param redeemScript Only overrides if there is no redeemScript in the input currently
+ */
+export function updateReplayProtectionUnspentToPsbt(
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/wallet/Unspent.ts:152]:
```suggestion
export function updatePsbtRedeemScript(
```

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]
few naming suggestions

File: modules/utxo-lib/src/bitgo/wallet/Unspent.ts [hunk 5]
```diff
@@ -120,129 +122,220 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
...
   });
-  if (isZcash) {
-    psbt.updateInput(psbt.inputCount - 1, { witnessUtxo: { script, value } });
-  } else {
-    psbt.updateInput(psbt.inputCount - 1, { nonWitnessUtxo: (u as UnspentWithPrevTx<bigint>).prevTx });
-  }
 }
 
-export function addWalletUnspentToPsbt(
+export function addReplayProtectionUnspentToPsbt(psbt: UtxoPsbt, u: Unspent<bigint>, redeemScript: Buffer): void {
```
[the remark this answers is not in the document]
Comment (davidkaplanbitgo) [reply modules/utxo-lib/src/bitgo/wallet/Unspent.ts:190]: Not sure if it is a good idea to change the name of this function in this PR - would have breaking changes

[the remark this answers is not in the document]
Comment (OttoAllmendinger) [reply modules/utxo-lib/src/bitgo/wallet/Unspent.ts:190]: good point

File: modules/utxo-lib/src/bitgo/wallet/Unspent.ts [hunk 6]
```diff
@@ -120,129 +130,232 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
...
+  const isZcash = getMainnet(psbt.network) === networks.zcash;
   if (!isUnspentWithPrevTx(u) && !isZcash) {
     throw new Error('Error, require previous tx to add to PSBT');
   }
+  if (isZcash && !input.witnessUtxo) {
+    const { script, value } = toPrevOutput(u, psbt.network);
+    psbt.updateInput(psbt.inputCount - 1, { witnessUtxo: { script, value } });
+  } else if (!isZcash && !input.nonWitnessUtxo) {
+    psbt.updateInput(psbt.inputCount - 1, { nonWitnessUtxo: (u as UnspentWithPrevTx<bigint>).prevTx });
+  }
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/wallet/Unspent.ts:181]: shouldn't you use `inputIndex` here?

File: modules/utxo-lib/test/bitgo/psbt/Psbt.ts [hunk 7]
```diff
@@ -340,6 +354,107 @@ describe('isPsbt', function () {
...
+  const componentsOnEachScriptType = {
+    p2sh: ['nonWitnessUtxo', 'redeemScript', 'bip32Derivation'],
+    p2shP2wsh: ['witnessUtxo', 'bip32Derivation', 'redeemScript', 'witnessScript'],
+    p2wsh: ['witnessUtxo', 'witnessScript', 'bip32Derivation'],
+    p2tr: ['witnessUtxo', 'tapLeafScript', 'tapBip32Derivation'],
+    p2trMusig2: ['witnessUtxo', 'tapBip32Derivation', 'tapInternalKey', 'tapMerkleRoot', 'unknownKeyVals'],
+    p2shP2pk: ['redeemScript', 'nonWitnessUtxo'],
+  };
+  inputScriptTypes.forEach((scriptType, i) => {
+    componentsOnEachScriptType[scriptType].forEach((inputComponent) => {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/bitgo/psbt/Psbt.ts:441]: nice test

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

**Review by OttoAllmendinger — APPROVED** [review]

**Review by saravanan7mani — APPROVED** [review]

**Pull request merged**